src/core/iter.rs

```diff
@@ -0,0 +1,98 @@
+//
+
+//! Iterators for DAG-like structures
+
+use std::iter::Rev;
+use std::vec::IntoIter;
+
+/// A DAG-like structure with a root,
+/// left children and right children per node,
+/// without directed cycles.
+///
+/// We use this trait to implement general DAG operations, like node traversal.
+pub trait DagLike: Sized {
+    /// Node in the DAG
+    type Node: Copy;
+    /// Additional data for internal computations
+    type Extra;
+
+    /// Return the DAG root, if the DAG is nonempty.
+    fn root(&self) -> Option<Self::Node>;
+
+    /// Return the left child of the given `node`.
+    fn left_of(&self, node: Self::Node) -> Option<Self::Node>;
+
+    /// Return the right child of the given `node`.
+    fn right_of(&self, node: Self::Node) -> Option<Self::Node>;
+
+    /// During post-order traversal in [`Self::get_post_order_stack`],
+    /// return whether `left` or `right` are a cross-reference to nodes
+    /// that appear earlier in the order.
+    ///
+    /// If `maybe_right = Some(right)`, then `right` is checked.
+    /// Otherwise, `left` is checked.
+    ///
+    /// Arguments `to_visit` and `extra` provide additional data.
+    fn is_cross_reference(
+        &self,
+        node: Self::Node,
+        left: Self::Node,
+        maybe_right: Option<Self::Node>,
+        to_visit: &[Self::Node],
+        extra: &Self::Extra,
+    ) -> bool;
+
+    /// Return additional data for internal computations.
+    fn get_extra(&self) -> Self::Extra;
+
+    /// Visit the DAG and return a stack that can be popped such that the nodes are in post order.
+    ///
+    /// Post order means, left children appear before right ones, and children appear before their parent.
+    /// Shared nodes appear only once at their leftmost position.
+    fn get_post_order_stack(&self) -> Vec<Self::Node> {
+        let mut to_visit = match self.root() {
+            Some(node) => vec![node],
+            None => return Vec::new(),
+        };
+        let mut visited = Vec::new();
+        let extra = self.get_extra();
+
+        while let Some(node) = to_visit.pop() {
+            visited.push(node);
+
+            if let Some(left) = self.left_of(node) {
+                if !self.is_cross_reference(node, left, None, &to_visit, &extra) {
+                    to_visit.push(left);
+                }
+
+                if let Some(right) = self.right_of(node) {
+                    if !self.is_cross_reference(node, left, Some(right), &to_visit, &extra) {
+                        to_visit.push(right);
+                    }
+                }
+            }
+        }
+
+        visited
+    }
+
+    /// Return an iterator that visits the DAG in post order.
+    ///
+    /// Post order means, left children appear before right ones, and children appear before their parent.
+    /// Shared nodes appear only once at their leftmost position.
+    fn iter_post_order(&self) -> Rev<IntoIter<Self::Node>> {
+        self.get_post_order_stack().into_iter().rev()
+    }
```

Comment:
In 925b8cbdb0be693e1fa8d93005eb87d2a9c54846:

Why is this split into two functions? I would create a custom iterator type which maintains the stack and just returns elements as they are requested. Also, importing `vec::IntoIter` is confusing because `IntoIter` is already a trait name which is included in the std prelude.

Reply:
By two functions, do you mean `into_iter()` and `rev()`? My goal was to avoid a new iterator type if the same functionality can be had using standard functions. Arguably, the return type of `iter_post_order()` turned into something nasty due to that. Is there no way around writing "duplicate" code?

Reply:
I mean `get_post_order_stack` and `iter_post_order`. Splitting them up still involves an unnecessary allocation and exposes the user to the reversed logic.

Reply:
I realised that having the stack method in a public interface is not ideal, so I moved the logic into a new post-order iterator type. That being said, I don't think there was any unnecessary allocation, unless I am missing something. The stack is necessary and needs to grow to its maximum size until we can pop from it.